### ISSUE TITLE
Numeric hypotheses have to be specified on the link scale?

### ISSUE THREAD

**Issue by andrewheiss** [vincentarelbundock/marginaleffects#1447]
(idk if this works better as a discussion instead of an issue—feel free to move it over there if so!)

In the ["Predictions" section of the bonus hypothesis tests chapter](https://marginaleffects.com/bonus/hypothesis.html#predictions), it currently shows this example of changing null hypotheses:

<img width="639" alt="Image" src="https://github.com/user-attachments/assets/f2d7cffe-a771-4c20-ad94-1cf4b4a2e380" />

---

Since {marginaleffects} reports values on the probability scale by default, the example currently seems to imply that specifying hypotheses is also done on the probability scale. Like this example as currently written:

```r
predictions(mod, newdata = "mean", hypothesis = .5)
```

```
 Estimate Pr(>|z|)   S  2.5 % 97.5 %  hp drat
    0.231   0.0343 4.9 0.0584  0.592 147  3.6

Type:  invlink(link) 
Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, hp, drat, am 
```

…seems like it should be testing if the predicted value is different from a null of 50%. However, 50% is safely inside the reported confidence interval, but the p-value is less than 0.05. Similarly, when using the default null of 0, 0 is safely out of the the reported confidence interval, but the p-value is greater than 0.05.

I've included a longer explanation below, but here's the tl;dr version: 

> [!IMPORTANT]
> The documentation might need to be updated to say that numeric hypotheses are specified in the link units, not the transformed units (probability here with logits; logs vs. counts for Poisson, etc.)
> 
> It might already say that in other parts of the documentation, but it doesn't in the "Hypothesis Tests (Bonus)" section and as currently written, it feels like `hypothesis = 0.5` implies 50% and not the actual 62% that it's testing.

(I could be wrong though!)

---

To illustrate this better, first I'll use base R functions to find the predicted value and confidence interval:

```r
preds_logit <- predict(
  mod,
  newdata = data.frame(hp = mean(mtcars$hp), drat = mean(mtcars$drat)),
  se.fit = TRUE,
  type = "link"
)

fit_logit <- preds_logit$fit
se_fit_logit <- preds_logit$se.fit

# Confidence interval on link/logit scale
z <- qnorm(1 - 0.05/2)
lower_logit <- fit_logit - (z * se_fit_logit)
upper_logit <- fit_logit + (z * se_fit_logit)

# Logit to probabilities
tibble(fit_logit, se_fit_logit, lower_logit, upper_logit) |> 
  mutate(across(
    c(fit_logit, lower_logit, upper_logit), 
    list(prob = \(x) plogis(x))
  ))
```

```
# A tibble: 1 × 7
  fit_logit se_fit_logit lower_logit upper_logit fit_logit_prob lower_logit_prob upper_logit_prob
      <dbl>        <dbl>       <dbl>       <dbl>          <dbl>            <dbl>            <dbl>
1     -1.20        0.805       -2.78       0.374          0.231           0.0584            0.592
```

Here we have the predicted value and confidence interval on both the link/logit scale and on the probability scale, and it matches what we get with `predictions()`:

```r
# These estimates, SEs, and confidence intervals are all identical to the manual
# calculations above
predictions(mod, newdata = "mean", type = "link")
```

```
 Estimate Std. Error    z Pr(>|z|)   S 2.5 % 97.5 %  hp drat
     -1.2      0.805 -1.5    0.135 2.9 -2.78  0.374 147  3.6

Type:  link 
Columns: rowid, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high, hp, drat, am 
```

```r
predictions(mod, newdata = "mean")
```

```
 Estimate Pr(>|z|)   S  2.5 % 97.5 %  hp drat
    0.231    0.135 2.9 0.0584  0.592 147  3.6

Type:  invlink(link) 
Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, hp, drat, am 
```

Interestingly the p-value is identical in both of these outputs (0.135). The hypothesis here is whether the **logit scale prediction** is 0. Here it is manually:

```r
# p-value for whether fit_logit == 0
tibble(fit_logit, se_fit_logit, lower_logit, upper_logit) |> 
  mutate(
    test_stat = (fit_logit - 0) / se_fit_logit,
    pvalue = 2 * pnorm(-abs(test_stat))
  )
```

```
# A tibble: 1 × 6
  fit_logit se_fit_logit lower_logit upper_logit test_stat pvalue
      <dbl>        <dbl>       <dbl>       <dbl>     <dbl>  <dbl>
1     -1.20        0.805       -2.78       0.374     -1.50  0.135
```

-1.2 isn't significantly different from 0 (p = 0.135). **But that 0 here doesn't mean 0%**. It actually means `plogis(0)`, or 50%:

```r
plogis(0)
```

```
[1] 0.5
```

So in this default output on the probability scale, the p-value is actually testing if 0.231 is equivalent to 50%, not 0%:

```r
predictions(mod, newdata = "mean", hypothesis = 0)
```

```
 Estimate Pr(>|z|)   S  2.5 % 97.5 %  hp drat
    0.231    0.135 2.9 0.0584  0.592 147  3.6

Type:  invlink(link) 
Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, hp, drat, am 
```

If we specify `hypothesis = 0.5`, the hypothesis test will also be on the link scale. Here's the manual p-value:

```r
# p-value for whether fit_logit == 0.5
tibble(fit_logit, se_fit_logit, lower_logit, upper_logit) |> 
  mutate(
    test_stat = (fit_logit - 0.5) / se_fit_logit,
    pvalue = 2 * pnorm(-abs(test_stat))
  )
```

```
# A tibble: 1 × 6
  fit_logit se_fit_logit lower_logit upper_logit test_stat pvalue
      <dbl>        <dbl>       <dbl>       <dbl>     <dbl>  <dbl>
1     -1.20        0.805       -2.78       0.374     -2.12 0.0343
```

That's 0.0343, which is significant, which means that -1.2 is likely not equivalent to 0.5. This matches what we get when working with `predictions(..., type = "link")`:

```r
predictions(mod, newdata = "mean", type = "link", hypothesis = 0.5)
```

```
 Estimate Std. Error     z Pr(>|z|)   S 2.5 % 97.5 %  hp drat
     -1.2      0.805 -2.12   0.0343 4.9 -2.78  0.374 147  3.6

Type:  link 
Columns: rowid, estimate, std.error, statistic, p.value, s.value, conf.low, conf.high, hp, drat, am 
```

But it's also the same p-value when working on the probability scale:

```r
predictions(mod, newdata = "mean", hypothesis = 0.5)
```

```
 Estimate Pr(>|z|)   S  2.5 % 97.5 %  hp drat
    0.231   0.0343 4.9 0.0584  0.592 147  3.6

Type:  invlink(link) 
Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, hp, drat, am
```

On the probability scale, 0.5 logits is actually `plogis(0.5)` or ≈62%:

```r
plogis(0.5)
```

```
[1] 0.6224593
```

So this output:

```r
predictions(mod, newdata = "mean", hypothesis = 0.5)
```

```
 Estimate Pr(>|z|)   S  2.5 % 97.5 %  hp drat
    0.231   0.0343 4.9 0.0584  0.592 147  3.6

Type:  invlink(link) 
Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, hp, drat, am 
```

…actually shows the p-value for whether 0.231 is equivalent to 0.62 (and it's not, since p < 0.05). 0.5 appears in the confidence interval; 0.62 does not.

---

So, if we want to test hypotheses on the probability scale, we need to specify the null value on the link scale:

```r
# Is the predicted value different from 50%?
predictions(mod, newdata = "mean", hypothesis = qlogis(0.5))

# or just 0, since qlogis(0.5) = 0
# predictions(mod, newdata = "mean", hypothesis = 0)
```

```
 Estimate Pr(>|z|)   S  2.5 % 97.5 %  hp drat
    0.231    0.135 2.9 0.0584  0.592 147  3.6

Type:  invlink(link) 
Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, hp, drat, am 
```

If we want to know if 0.231 is significantly different from 0, we can't use `qlogis(0)` because that's infinity:

```r
qlogis(0)
```

```
[1] -Inf
```

But we can see if it's smaller than some small number, like 1%:

```r
# Is the predicted value different from 1%?
predictions(mod, newdata = "mean", hypothesis = qlogis(0.01))
```

```
 Estimate Pr(>|z|)    S  2.5 % 97.5 %  hp drat
    0.231   <0.001 15.3 0.0584  0.592 147  3.6

Type:  invlink(link) 
Columns: rowid, estimate, p.value, s.value, conf.low, conf.high, hp, drat, am 
```

**Comment by andrewheiss** [comment]
The official documentation does say that numeric null hypotheses are on whatever scale is used before `transform`, but the `transform` argument by default is NULL, which makes it feel like maybe a separate process than what happens when using `type = "invlink(link)"`

<img width="707" alt="Image" src="https://github.com/user-attachments/assets/075d4969-c081-4821-9571-6dfeb685d514" />

**Comment by vincentarelbundock** [comment]
What happens if you manually specify `type="response"`?

**Comment by andrewheiss** [comment]
Huh, strange-ish things. It seems to work, but it leads to negative CIs:

``` r
library(marginaleffects)

mod <- glm(am ~ hp + drat, data = mtcars, family = binomial)

# Probability-scale estimate and CI, but p-value is from 
# "check if logit-scale estimate is 0, or 50%" test
predictions(mod, newdata = "mean")
#> 
#>  Estimate Pr(>|z|)   S  2.5 % 97.5 %
#>     0.231    0.135 2.9 0.0584  0.592
#> 
#> Type:  invlink(link)

# Same output
predictions(mod, newdata = "mean", type = "invlink(link)")
#> 
#>  Estimate Pr(>|z|)   S  2.5 % 97.5 %
#>     0.231    0.135 2.9 0.0584  0.592
#> 
#> Type:  invlink(link)

# Probability-scale estimate, but CI is negative now, and p-value is from some
# sort of test? Probably if the probability-scale estimate is 0%, but it's not
# significant here because the CI is negative
predictions(mod, newdata = "mean", type = "response")
#> 
#>  Estimate Std. Error    z Pr(>|z|)   S   2.5 % 97.5 %
#>     0.231      0.143 1.62    0.106 3.2 -0.0491  0.511
#> 
#> Type:  response


# Probability-scale estimate and CI, but p-value is from 
# "check if logit-scale estimate is 0.5, or 62%" test
predictions(mod, newdata = "mean", hypothesis = 0.5)
#> 
#>  Estimate Pr(>|z|)   S  2.5 % 97.5 %
#>     0.231   0.0343 4.9 0.0584  0.592
#> 
#> Type:  invlink(link)

# Same output
predictions(mod, newdata = "mean", type = "invlink(link)", hypothesis = 0.5)
#> 
#>  Estimate Pr(>|z|)   S  2.5 % 97.5 %
#>     0.231   0.0343 4.9 0.0584  0.592
#> 
#> Type:  invlink(link)

# Probability-scale estimate, but CI is negative now, though the p-value is from
# test if estimate is equivalent to 50%
predictions(mod, newdata = "mean", type = "response", hypothesis = 0.5)
#> 
#>  Estimate Std. Error     z Pr(>|z|)   S   2.5 % 97.5 %
#>     0.231      0.143 -1.88   0.0598 4.1 -0.0491  0.511
#> 
#> Type:  response
```

**Comment by andrewheiss** [comment]
That means that these two tests, which should theoretically be the same thing (unless I'm doing something or thinking about something wrong??), give different answers:

``` r
library(marginaleffects)

mod <- glm(am ~ hp + drat, data = mtcars, family = binomial)
```

Is the prediction different from 50% (using 0 on the logit scale)? No.

```r
predictions(mod, newdata = "mean", hypothesis = 0, type = "link")
#> 
#>  Estimate Std. Error    z Pr(>|z|)   S 2.5 % 97.5 %
#>      -1.2      0.805 -1.5    0.135 2.9 -2.78  0.374
#> 
#> Type:  link
```

Is the prediction different from 50% (using 50% on the response scale)? Yes.

```r
predictions(mod, newdata = "mean", hypothesis = 0.5, type = "response")
#> 
#>  Estimate Std. Error     z Pr(>|z|)   S   2.5 % 97.5 %
#>     0.231      0.143 -1.88   0.0598 4.1 -0.0491  0.511
#> 
#> Type:  response
```

**Comment by vincentarelbundock** [comment]
### `type="link"`

Does inference (se, t, p, ci) directly on the link scale. The disadvantage is that the scale is hard to interpret substantively.

### `type="response"`

Does inference (se, t, p, ci) directly on the response scale. The disadvantage is that you can get estimates outside the "natural" bounds of the response.

### `type="invlink(link)`

Does inference (se, t, p, ci) directly on the link scale. Then, it inserts a function in the `transform` argument to "back-transform" the link scale predictions to the response scale. 

* Upside: predictions cannot go outside natural bounds
* Downside: we must do inference on the link scale, which complicates the specification of the null hypothesis

From an implementation perspective, note that the `transform` argument we use to backtransform is always executed at the very end. https://marginaleffects.com/man/r/predictions.html#order-of-operations

I'd be very interested in suggestions on how to document this more clearly.

**Comment by andrewheiss** [comment]
Oh ok awesome, that all makes perfect sense!

The main reason I stumbled on this was that someone DMed me about it on Bluesky bc they were working through the [bonus hypothesis test section here](https://marginaleffects.com/bonus/hypothesis.html#predictions) and assumed that the 0.5 referred to probability scale stuff, since there's no indication anywhere else in the output that inferential things are based on the link scale

<img width="625" alt="Image" src="https://github.com/user-attachments/assets/d10a77bd-e9d6-4c82-a064-919302cc3339" />

So an easy quick fix is probably just to change that example to 

```
predictions(mod, newdata = "mean", type = "response")
predictions(mod, newdata = "mean", hypothesis = .5, type = "response")
```

or something similar?

I don't know how verbose you'd want to get with messages or notes, but `parameters::model_parameters()` includes extra details about the scale of coefficients and how p-values are calculated:

```
> parameters::model_parameters(mod)
Parameter   | Log-Odds |       SE |           95% CI |     z |     p
--------------------------------------------------------------------
(Intercept) |   -29.08 |    12.42 | [-63.19, -11.57] | -2.34 | 0.019
hp          |     0.01 | 9.33e-03 | [ -0.01,   0.03] |  1.16 | 0.247
drat        |     7.31 |     3.05 | [  3.04,  15.73] |  2.40 | 0.016

Uncertainty intervals (profile-likelihood) and p-values (two-tailed) computed using a Wald z-distribution approximation.

The model has a log- or logit-link. Consider using `exponentiate = TRUE` to interpret coefficients as ratios.
```

Something similar could maybe work here, but would require like a whole separate messaging mechanism, and would lead to messier output (I find myself always having to specify `verbose = FALSE` with `model_parameters()` and often find it too chatty), so it's probably not a great actual solution.

Or maybe adding something to one of the book chapters on hypothesis tests with models that use different link functions?

**Comment by ngreifer** [comment]
One strategy would be to automatically transform the supplied null value into the link scale for inference. That way the user doesn't have to think about the link and can just test the hypothesis on the natural scale. You would have to trigger this only when `type = invlink(link)`. I agree the documentation should be updated to use a more illustrative example.

**Comment by vincentarelbundock** [comment]
That would work when `hypothesis` is a single number, but it is more complicated to handle internally when users supply arbitrary string equations like `hypothesis="b1 / b2 = exp(2)"`

The two cases should be handled the same way, so if I can't figure out a low-maintenance implementation for the more complicated case, I think the solution might be to document this better and perhaps raise a once-per-session warning.

**Comment by vincentarelbundock** [comment]
Thanks both for the useful report and discussion.

Just merged a 3-pronged resolution:

1. Man page of the `hypothesis` argument now mentions that when `type='invlink(link)'`, p values are computed on the link scale before back transforming.
2. Raise a warning once per session when this happens. 
3. Added information to the vignette linked above
